File: sdk/python/stele/producer.py
"""High-level Producer class for stele.

Wraps :class:`stele.PrivateKey` + :class:`stele.SteleClient` so the
caller does ``producer.log(...)`` and ``producer.enroll(...)`` without
touching envelope canonicalisation or HTTP details.
"""

from __future__ import annotations

import ssl
from base64 import b64decode
from typing import Any, Dict, Optional

from .client import SteleClient, b64decode_str, b64encode_bytes
from .envelope import Envelope, new_envelope
from .keys import PrivateKey


class ProducerError(RuntimeError):
    """Raised when the producer's high-level operation fails before HTTP."""


class Producer:
    """A stele producer: signs envelopes, talks to the operator.

    A producer is bound to:
      - a stable ``id`` (the operator registers this)
      - an Ed25519 :class:`PrivateKey`
      - a single operator (``server``)

    Examples
    --------
    >>> from stele import Producer, generate_key
    >>> priv = generate_key()
    >>> p = Producer(id="alice", private_key=priv,
    ...              server="http://localhost:8080")
    >>> p.enroll(scope="logs:audit", validity_seconds=86400 * 90)
    {'enrollment': ...}
    >>> p.log(source="my.app", data=b"first entry")['entry']['index']
    0
    """

    def __init__(
        self,
        *,
        id: str,
        private_key: PrivateKey,
        server: str,
        attestation_type: str = "software",
        ssl_context: Optional[ssl.SSLContext] = None,
        timeout: float = 15.0,
    ) -> None:
        if not id:
            raise ValueError("Producer: id required")
        self.id = id
        self.private_key = private_key
        self.attestation_type = attestation_type
        self.client = SteleClient(
            base_url=server,
            timeout=timeout,
            ssl_context=ssl_context,
        )

    # ---------- enrollment ----------

    def enroll(
        self,
        *,
        scope: str = "",
        description: str = "",
        validity_seconds: int = 0,
    ) -> Dict[str, Any]:
        """Run the two-step proof-of-possession enrollment ceremony.

        Step 1 (``POST /api/v0/enrollments/begin``) tells the operator
        the producer's identity + pubkey and gets back a challenge.

        Step 2 (``POST /api/v0/enrollments/confirm``) signs the
        challenge with the private key and sends the signature. On
        success the producer is registered + enrolled with a
        signature from the operator's chain key over the producer's
        identity AND a proof from the producer that they control the
        key.

        Returns the server's confirmation response (parsed JSON).
        Raises :class:`HTTPError` on any non-2xx response.
        """
        begin_req: Dict[str, Any] = {
            "id": self.id,
            "public_key": b64encode_bytes(self.private_key.public_bytes()),
            "attestation_type": self.attestation_type,
        }
        if scope:
            begin_req["scope"] = scope
        if description:
            begin_req["description"] = description
        if validity_seconds > 0:
            begin_req["validity_seconds"] = int(validity_seconds)

        begin = self.client.post("/api/v0/enrollments/begin", begin_req)
        challenge_id = begin["challenge_id"]
        challenge = b64decode_str(begin["challenge_bytes"])

        signature = self.private_key.sign(challenge)
        confirm_req = {
            "challenge_id": challenge_id,
            "signature": b64encode_bytes(signature),
        }
        return self.client.post("/api/v0/enrollments/confirm", confirm_req)

    # ---------- logging ----------

    def log(
        self,
        *,
        source: str,
        data: bytes,
        honeypot: bool = False,
    ) -> Dict[str, Any]:
        """Sign + submit one envelope; return the operator's response.

        ``source`` is a free-form string identifying where this entry
        came from (e.g. a file path, a service name). ``data`` is the
        raw bytes you're recording — stele does not inspect them.

        ``honeypot=True`` marks the entry as a canary on the operator
        side; only enable on entries you want to be unread and which
        will trigger the operator's honeylog on access.

        Returns the parsed JSON response from ``/api/v0/append``,
        which includes the operator-assigned ``index`` under
        ``response["entry"]["index"]``.
        """
        env = new_envelope(
            producer_id=self.id,
            public_key=self.private_key.public_bytes(),
            source=source,
            data=data,
            attestation_type=self.attestation_type,
        )
        env.sign(self.private_key)
        req = {"envelope": env.to_dict(), "honeypot": honeypot}
        return self.client.post("/api/v0/append", req)

    # ---------- introspection ----------

    def server_pubkey(self) -> bytes:
        """Fetch the operator's root pubkey (auditor trust anchor)."""
        resp = self.client.get("/api/v0/pubkey")
        return b64decode_str(resp["root_public_key"])

    def size(self) -> int:
        """Current size of the operator's log."""
        return int(self.client.get("/api/v0/size")["size"])

File: sdk/typescript/src/producer.ts
// High-level Producer class — wraps PrivateKey + SteleClient so the
// caller does `await producer.log({...})` and `await producer.enroll({...})`
// without touching envelope canonicalisation or HTTP details.

import {
  Envelope,
  TYPE_SOFTWARE,
  b64,
  b64decode,
  canonicalBytes,
  envelopeToWire,
} from "./envelope.js";
import { SteleClient, SteleClientOptions } from "./client.js";
import { PrivateKey } from "./keys.js";

export interface ProducerOptions {
  id: string;
  privateKey: PrivateKey;
  server: string;
  attestationType?: string;
  timeoutMs?: number;
  fetchImpl?: typeof fetch;
}

export interface EnrollOptions {
  scope?: string;
  description?: string;
  validitySeconds?: number;
}

export interface LogOptions {
  source: string;
  data: Uint8Array | string;
  honeypot?: boolean;
}

interface BeginResponse {
  challenge_id: string;
  challenge_bytes: string; // base64
  expires_at_ns: number;
}

export class Producer {
  readonly id: string;
  readonly privateKey: PrivateKey;
  readonly attestationType: string;
  readonly client: SteleClient;

  constructor(opts: ProducerOptions) {
    if (!opts.id) throw new Error("Producer: id required");
    if (!opts.privateKey) throw new Error("Producer: privateKey required");
    this.id = opts.id;
    this.privateKey = opts.privateKey;
    this.attestationType = opts.attestationType ?? TYPE_SOFTWARE;
    this.client = new SteleClient({
      baseUrl: opts.server,
      timeoutMs: opts.timeoutMs,
      fetchImpl: opts.fetchImpl,
    } satisfies SteleClientOptions);
  }

  /** Two-step proof-of-possession enrollment ceremony. */
  async enroll(opts: EnrollOptions = {}): Promise<unknown> {
    const beginReq: Record<string, unknown> = {
      id: this.id,
      public_key: b64(this.privateKey.publicBytes()),
      attestation_type: this.attestationType,
    };
    if (opts.scope) beginReq.scope = opts.scope;
    if (opts.description) beginReq.description = opts.description;
    if (opts.validitySeconds && opts.validitySeconds > 0) {
      beginReq.validity_seconds = opts.validitySeconds;
    }

    const begin = await this.client.post<BeginResponse>(
      "/api/v0/enrollments/begin",
      beginReq,
    );
    const challenge = b64decode(begin.challenge_bytes);
    const signature = await this.privateKey.sign(challenge);

    return this.client.post("/api/v0/enrollments/confirm", {
      challenge_id: begin.challenge_id,
      signature: b64(signature),
    });
  }

  /** Sign + submit one envelope; return the operator's response. */
  async log(opts: LogOptions): Promise<unknown> {
    const data =
      typeof opts.data === "string" ? new TextEncoder().encode(opts.data) : opts.data;

    const env: Envelope = {
      producerId: this.id,
      timeNanos: BigInt(Date.now()) * 1_000_000n,
      source: opts.source,
      data,
      publicKey: this.privateKey.publicBytes(),
      attestationType: this.attestationType,
    };

    const canonical = canonicalBytes({
      producerId: env.producerId,
      timeNanos: env.timeNanos,
      source: env.source,
      data: env.data,
      publicKey: env.publicKey,
      attestationType: env.attestationType,
    });
    env.signature = await this.privateKey.sign(canonical);

    return this.client.post("/api/v0/append", {
      envelope: envelopeToWire(env),
      honeypot: opts.honeypot ?? false,
    });
  }

  /** Fetch the operator's root pubkey (auditor trust anchor). */
  async serverPubkey(): Promise<Uint8Array> {
    const resp = await this.client.get<{ root_public_key: string }>(
      "/api/v0/pubkey",
    );
    return b64decode(resp.root_public_key);
  }

  /** Current size of the operator's log. */
  async size(): Promise<number> {
    const resp = await this.client.get<{ size: number }>("/api/v0/size");
    return resp.size;
  }
}
